### PR TITLE
bugfix: allow quoted exemplar keys in openmetrics text format

### LINKED ISSUE #15259: UTF-8: scrape fails with quoted "utf8" key in exemplar
### What did you do?

Expose an a metric with exemplar from client_go (via opentelemetry), where one of the labels features a non-traditional character in the label (e.g. `http.scheme`),

### What did you expect to see?

This is exported in sccraping as:
```
metric_total{n="1"} 0 # {"net.http"="abc"} 1.0
```
Prometheus then fails to scrape this due to the presence of the quote:
```
 expected label name, got "\"" ("INVALID") while parsing: "metric_total{n=\"1\"} 0 # {\""
```

### What did you see instead? Under which circumstances?

Prometheus should successfully scrape the quote label name.

### System information

Linux 6.11.5-arch1-1 x86_64

### Prometheus version

```text
tested on 
commit: dd4eb4590dc7f4163afb452fd3f2f189caf2d20f

discovered via opentelemetry collector using 

github.com/prometheus/prometheus v0.54.1
```
```


### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by tcolgate**
Work in progress, I'm not sure if this is sufficient (should the regular prometheus text format include this too?)

Fixes #15259

### CODE REVIEW

**Comment by ArthurSens** [comment]
We have a test dedicated for UTF a few lines below called `TestUTF8OpenMetricsParse`, could you move it there? About prometheus parser, yes! We also have the same test over there :)

About the correctness of the code, I'm not particular familiar with the lexer. Maybe @ywwg or @bboreham ?

**Comment by tcolgate** [comment]
> We have a test dedicated for UTF8 a few lines below called `TestUTF8OpenMetricsParse`, could you move it there?

Moved :+1: 

> About prometheus parser, yes! We also have the same test in promparse_test.go :)

Will take a look

**Comment by ArthurSens** [comment]
> Having a quick look, the prom parser include a constant for the exemplar state, but doesn't actually seem to parse them. From memory, that is correct, I think you have to enable openmetrics in the promhttp handler to export exemplars, I think the definition is specific to OM and they aren't supported by a regular, non-OM scrape. This may be wrong, and I may be misunderstanding how the parser actually ends up working.

Good point, I believe you're correct!

**Comment by ArthurSens** [comment]
@tcolgate , could you rebase your PR? The CI was fixed on main already. @ywwg, could you take a look once you find some time?

**Review by krajorama — APPROVED** [review]
LGTM

File: model/textparse/openmetricsparse_test.go [hunk 1]
```diff
@@ -486,7 +486,9 @@ func TestUTF8OpenMetricsParse(t *testing.T) {
 {"http.status",q="0.9",a="b"} 8.3835e-05
 {q="0.9","http.status",a="b"} 8.3835e-05
 {"go.gc_duration_seconds_sum"} 0.004304266
-{"Heizölrückstoßabdämpfung 10€ metric with \"interesting\" {character\nchoices}","strange©™\n'quoted' \"name\""="6"} 10.0`
+{"Heizölrückstoßabdämpfung 10€ metric with \"interesting\" {character\nchoices}","strange©™\n'quoted' \"name\""="6"} 10.0
```
Comment (bboreham) [inline model/textparse/openmetricsparse_test.go:489]: Note if the closing quote here was on the next line, we wouldn't get a diff on this line.
Suggest you break the cycle by changing your line 491 to do this.

Comment (tcolgate) [reply model/textparse/openmetricsparse_test.go:489]: Unfortunately that actually breaks on parsing (`expected a valid start token, got "\n" ("INVALID") while parsing: "\n"`), so it'd also require a trim of the string before parsing. I can include that, but the extra trim might look a bit odd to future reviewers.

Comment (tcolgate) [reply model/textparse/openmetricsparse_test.go:489]: yes, having a newline before the closing quote causes the parser to attempt to parse another line and hit EOF straight away, which raises the above error.

**Comment by tcolgate** [comment]
Is merging of this contingent on the OpenMetrics v2 effort?

**Comment by bwplotka** [comment]
Yes, we need to officially put that in spec, but if it's "additive" feature, it could be merged to Prometheus same as [other UTF-8 changes](https://github.com/prometheus/prometheus/commit/a28d7865adf6542cbac3b908124ae07846a0d4a8).

**Review by bwplotka — APPROVED** [review]

**Pull request merged**